src/LupaClient.php

```diff
@@ -0,0 +1,191 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LupaSearch;
+
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\Exception\GuzzleException;
+use GuzzleHttp\Utils;
+use LupaSearch\Exceptions\AuthenticationException;
+use LupaSearch\Exceptions\MissingCredentialsException;
+use LupaSearch\Exceptions\TooManyRetriesException;
+use LupaSearch\Factories\HttpClientFactory;
+use LupaSearch\Factories\HttpClientFactoryInterface;
+use LupaSearch\Utils\JwtUtils;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Psr7\Request;
+
+class LupaClient implements LupaClientInterface
+{
+    /**
+     * @var HttpClientFactoryInterface|null
+     */
+    private $httpClientFactory;
+
+    /**
+     * @var ClientInterface|null
+     */
+    private $httpClient;
+
+    /**
+     * @var string|null
+     */
+    private $jwtToken;
+
+    /**
+     * @var string|null
+     */
+    private $email;
+
+    /**
+     * @var string|null
+     */
+    private $password;
+
+    public function __construct(HttpClientFactoryInterface $httpClientFactory = null)
+    {
+        $this->httpClientFactory = $httpClientFactory ?? new HttpClientFactory();
```

Comment:
why create a factory and use it only once? Is't it better `$this->httpClient = ($httpClientFactory ?? new HttpClientFactory())->create()` approach?